core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -369,6 +369,31 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
+  /**
+   * try to complete produce, fetch and delete requests if the HW of partition is incremented. Otherwise, we try to complete
+   * only delayed fetch requests.
+   *
+   * Noted that this method may hold a lot of group lock so the caller should NOT hold any group lock
+   * in order to avoid deadlock
+   * @param topicPartitions topic partition and leaderHWIncremented
+   */
+  private[this] def completeDelayedRequests(topicPartitions: Map[TopicPartition, Boolean]): Unit =
+    topicPartitions.foreach {
+      case (tp, leaderHWIncremented) =>
+        if (leaderHWIncremented) groupManager.replicaManager.completeDelayedRequests(tp)
+        else groupManager.replicaManager.completeDelayedFetchRequests(tp)
+    }
+
+  /**
+   * complete the delayed join requests associated to input group keys.
+   *
+   * Noted that this method may hold a lot of group lock so the caller should NOT hold any group lock
+   * in order to avoid deadlock
+   * @param groupKeys group keys to complete
+   */
+  private[this] def completeDelayedJoinRequests(groupKeys: Set[GroupKey]): Unit =
+    groupKeys.foreach(joinPurgatory.checkAndComplete)
```

Comment:
> Does the group purgatory suffer from the same deadlock potential? If we call checkAndComplete for a group "foo," I don't think we would attempt completion for any other group.

you are right. It requires lock for group "foo" only. But the potential deadlock I tried to avoid/describe is that the caller has held a lock of group_a and then it tried to complete delayed request for group_b. It is possible to cause deadlock as it requires the lock of group_b to completing delayed request for group_b.

That way the comment says the caller should NOT hold any group lock.

Reply:
The join/heartbeat purgatories are a little different from the produce purgatory. The key is based on the groupId, so when we complete an operation for group "foo," we won't complete for group "bar" incidentally. At least that is my understanding. If you look at `DelayedOperationPurgatory.checkAndComplete`, we only complete watchers for the passed key.

Reply:
> so when we complete an operation for group "foo," we won't complete for group "bar" incidentally. At least that is my understanding. If you look at DelayedOperationPurgatory.checkAndComplete, we only complete watchers for the passed key.

You are totally right. My above comment is not clear. I added the comment to remind developers following code is dangerous.
```scala
lock group_a
completeDelayedJoinRequests for group_b (this line may produce deadlock if there is another thread which is holding lock of group_b and prepare to complete requests for group_a)
release group_a
```
Of course, the above code is nonexistent currently. I'm just worry that the deadlock is easy to produce in the future if we don't notice the lock issue.

Reply:
> so when we complete an operation for group "foo," we won't complete for group "bar"

In fact, there is a example of deadlock in join purgatories. 

```GroupCoordinator#onCompleteJoin``` called by ```DelayedJoin#onComplete``` is possible to call ```storeGroup``` to append records to __consumer_offsets-{**partitionFor(group.groupId)**}. If there are groups related to same partition, ```GroupCoordinator#onCompleteJoin``` which is holding a group lock will require locks for other groups.